### PR TITLE
Convergent encryption support

### PR DESCRIPTION
**Pull request by finalwharf**
What does this PR do?
--------------------
* Adds support for convergent encryption

Where should the reviewer start?
-------------------------------
* `lib/vault/rails.rb`
* `lib/vault/encrypted_model.rb`

Any background context you want to provide?
------------------------------------------
* Vault supports convergent encryption since v0.6.1, but this gem
does not take advantage of this functionality.

/cc @popovm @bliof please review

### CODE REVIEW

File: lib/vault/rails.rb [hunk 1]
```diff
@@ -27,6 +27,10 @@ module Rails
     DEV_WARNING = "[vault-rails] Using in-memory cipher - this is not secure " \
       "and should never be used in production-like environments!".freeze
 
+    CONVERGENT_ENCRYPTION_CONTEXT = Base64.strict_encode64(
+      ENV.fetch('VAULT_CONVERGENT_ENCRYPTION_CONTEXT', 'convergent-encryption-context')
```
Comment (popovm) [inline lib/vault/rails.rb]: Let's remove the default value here

File: lib/vault/rails.rb [hunk 2]
```diff
@@ -30,7 +30,7 @@ module Rails
     class << self
       # API client object based off the configured options in {Configurable}.
       #
-      # @return [Vault::Client]
+      # @return [Vault::Client] for use with convergent encryption
```
Comment (bliof) [inline lib/vault/rails.rb]: You can drop addition, you can use it without the convergent encryption

File: lib/vault/rails.rb [hunk 3]
```diff
@@ -143,49 +143,81 @@ def serializer_for(key)
...
 
         return nil if plaintext.nil?
 
         cipher = OpenSSL::Cipher::AES.new(128, :CBC)
         cipher.encrypt
         cipher.key = memory_key_for(path, key)
-        return Base64.strict_encode64(cipher.update(plaintext) + cipher.final)
+
+        iv = if convergent
+               cipher.iv = Vault::Rails.convergent_encryption_context.first(16)
```
Comment (bliof) [inline lib/vault/rails.rb]: We've talked that these `memory` methods are used only for development and testing, right?

File: lib/vault/rails.rb [hunk 4]
```diff
@@ -143,49 +143,81 @@ def serializer_for(key)
...
       end
 
       # Perform in-memory decryption. This is useful for testing and development.
-      def memory_decrypt(path, key, ciphertext, client)
+      def memory_decrypt(path, key, ciphertext, convergent)
         log_warning(DEV_WARNING) if self.in_memory_warnings_enabled?
 
         return nil if ciphertext.nil?
 
         cipher = OpenSSL::Cipher::AES.new(128, :CBC)
```
Comment (bliof) [inline lib/vault/rails.rb]: Does it make sense to try to use one of the ciphers in https://www.vaultproject.io/docs/secrets/transit/index.html#key-types for convergent encryption?

File: lib/vault/rails.rb (hunk 3 above)
Comment (finalwharf) [reply lib/vault/rails.rb]: Thats right! (btw and the size of the IV for AES::CBC is exactly 16 bytes)

File: lib/vault/rails.rb (hunk 2 above)
Comment (finalwharf) [reply lib/vault/rails.rb]: Accidental copy/paste

File: lib/vault/rails.rb (hunk 4 above)
Comment (bliof) [reply lib/vault/rails.rb]: Ahh just noticed the `log_warning(DEV_WARNING)` 👍

**Comment by finalwharf** [comment]
Some context: Vault::Rails has two protected methods - `memory_encrypt` and `memory_decrypt`, that is uses for encryption in development and testing environments instead of accessing Vault. The two methods are from the official repo, and use AES 128 CBC without an Initialization Vector.

Both using the same IV and no IV in the first place poses the same security issues. The reason for using the same IV is to make the gem behave the same both in development and production as the current PR adds support for convergent encryption for production as well.

The way that the VAULT_CONVERGENT_ENCRYPTION_CONTEXT is used is not the same in dev and prod. In development I use it as an IV (to keep things simple), but in prod, Vault uses it for key derivation.

Again, memory encryption is used only when Vault is not enabled (dev and test envs).

File: lib/vault/rails.rb [hunk 5]
```diff
@@ -143,49 +143,81 @@ def serializer_for(key)
       protected
 
       # Perform in-memory encryption. This is useful for testing and development.
-      def memory_encrypt(path, key, plaintext, client)
+      def memory_encrypt(path, key, plaintext, convergent)
```
Comment (bliof) [inline lib/vault/rails.rb]: This change most probably won't pass to `hashicorp/vault-rails`. It looks like they want to keep the same method signature.

(not a blocker, I would even change that protected to a private 🙊 )

**Comment by subodhsawant** [comment]
@finalwharf @bliof - RE: memory_encrypt / memory_decrypt / AES CBC 128 / static IV - this is fine from a security perspective as long as it is used for development and testing

File: spec/unit/rails_spec.rb [hunk 6]
```diff
@@ -20,4 +20,122 @@
       }
     end
   end
+
+  context 'when using convergent encryption' do
+    before :each do
+        allow(Vault::Rails).to receive(:enabled?).and_return(true)
```
Comment (bliof) [inline spec/unit/rails_spec.rb]: ✂️

File: spec/unit/rails_spec.rb [hunk 7]
```diff
@@ -20,4 +20,122 @@
...
+        allow(Vault::Rails).to receive(:convergent_encryption_context).and_return('a' * 16)
+    end
+
+    describe '.vault_encrypt' do
+      it 'sends the correct paramaters to vault client' do
+
```
Comment (bliof) [inline spec/unit/rails_spec.rb]: ✂️

File: spec/unit/rails_spec.rb [hunk 8]
```diff
@@ -20,4 +20,122 @@
...
+          context: Base64.strict_encode64(Vault::Rails.convergent_encryption_context),
+          convergent_encryption: true,
+          derived: true
+        }
+
+        expect(Vault::Rails.client.logical).to receive(:write).
+          with(expected_route, expected_options).
+          and_return(spy('Vault::Secret'))
+
+        Vault::Rails.send(:vault_encrypt, 'path', 'key', 'plaintext', Vault::Rails.client, true)
```
Comment (bliof) [inline spec/unit/rails_spec.rb]: Please use the public methods to test this

File: spec/unit/rails_spec.rb [hunk 9]
```diff
@@ -20,4 +20,117 @@
...
+      it 'calls memory_encrypt' do
+        allow(Vault::Rails).to receive(:enabled?).and_return(false)
+
+        expect(Vault::Rails).to receive(:memory_encrypt).with('path', 'key', 'plaintext', true).and_return('val')
+        Vault::Rails.encrypt('path', 'key', 'plaintext', nil, true)
+      end
+    end
+  end
+
+  describe '.decrypt' do
```
Comment (bliof) [inline spec/unit/rails_spec.rb]: tbo you can delete this and the one above

**Pull request merged**